lib/logger/lib/logger/backends/console.ex

```diff
@@ -97,4 +166,49 @@ defmodule Logger.Backends.Console do
   defp color_event(data, level, %{enabled: true} = colors) do
     [IO.ANSI.format_fragment(Map.fetch!(colors, level), true), data | IO.ANSI.reset]
   end
+
+  defp log_buffer(%{buffer_size: 0, buffer: []} = state), do: state
+
+  defp log_buffer(state) do
+    %{device: device, buffer: buffer} = state
+    %{state | ref: async_io(device, buffer), buffer: [], buffer_size: 0,
+      output: buffer}
+  end
+
+  defp handle_io_reply(:ok, %{ref: ref} = state) do
+    Process.demonitor(ref, [:flush])
+    log_buffer(%{state | ref: nil, output: nil})
+  end
+
+  defp handle_io_reply({:error, {:put_chars, :unicode, _} = error}, state) do
+    retry_log(error, state)
+  end
+
+  defp handle_io_reply({:error, :put_chars}, %{output: output} = state) do
+    retry_log({:put_chars, :unicode, output}, state)
+  end
+
+  defp handle_io_reply({:error, error}, _) do
+    raise "failure while logging console messages: " <> inspect(error)
+  end
+
+  defp retry_log(error, %{device: device, ref: ref, output: dirty} = state) do
+    Process.demonitor(ref, [:flush])
+    case :unicode.characters_to_binary(dirty) do
+      {_, good, bad} ->
+        clean = [good | Logger.Formatter.prune(bad)]
+        %{state | ref: async_io(device, clean), output: clean}
+      _ ->
+        # A well behaved IO device should not error on good data
+        raise "failure while logging consoles messages: " <> inspect(error)
```

Comment:
Instead of trying to relog just raise instead. This is simpler and I think log has no chance of success if we tried. Also a logging a batch may look strange if we tried.